File: Prog. Concu y Paralela/Tarea 4/Monitores/TiendaDekkerMonitor.java
package Monitores;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

public class TiendaDekkerMonitor {
    
    public static class MonitorDekker {
        private int turno = 0;
        
        public synchronized void entrar(int hiloId) {
            while (turno != hiloId) {
                try {
                    wait();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        }

        public synchronized void salir(int hiloId) {
            turno = 1 - hiloId;
            notifyAll();
        }
    }

    public static class Tienda {
        private HashMap<String, Integer> stock = new HashMap<>();
        ArrayList<String> prendas;

        public Tienda() {
            stock.put("Camisa Azul chica", 2);
            stock.put("Pantalon verde mediano", 3);
            stock.put("Tennis negros chicos", 5);
            prendas = new ArrayList<>(stock.keySet());
        }

        public String[] comprar(int numPrenda) {
            String prenda = prendas.get(numPrenda);
            Integer cantidad = stock.get(prenda);

            if (cantidad == 0)
                return new String[]{prenda, null};

            stock.put(prenda, cantidad - 1);
            return new String[]{prenda, String.valueOf(cantidad - 1)};
        }

        public Integer getSize() { return stock.size(); }

        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder();
            sb.append("Prenda\tCantidad disponible\n");
            stock.forEach(
                    (k, v) -> sb.append(k).append(":\t").append(v).append("\n"));
            return sb.toString();
        }
    }

    public static class Cliente extends Thread {
        private final MonitorDekker monitor;
        private final String nombre;
        private final Tienda tienda;
        private final int hiloId;

        public Cliente(String nombre, Tienda tienda, int hiloId, MonitorDekker monitor) {
            this.nombre = nombre;
            this.tienda = tienda;
            this.hiloId = hiloId;
            this.monitor = monitor;
        }

        @Override
        public void run() {
            int random = new Random().nextInt(tienda.getSize());

            monitor.entrar(hiloId);
            try {
                //SECCION CRITICA
                String[] resultado = tienda.comprar(random);
                //SECCION CRITICA

                if (resultado[1] == null)
                    System.err.println(nombre + " quiso comprar " + resultado[0] + 
                            " pero ya no hay existencias");
                else
                    System.out.println("- " + nombre + " ha comprado: " + resultado[0] +
                            " - Existencias restantes: " + resultado[1]);
            } finally {
                monitor.salir(hiloId);
            }
        }
    }

    public static void main(String[] args) throws InterruptedException {
        MonitorDekker monitor = new MonitorDekker();
        Thread[] hilos = new Thread[2];
        Tienda inst = new Tienda();
       
        System.out.println(inst.toString());
        hilos[0] = new Cliente("Pedro", inst, 0, monitor);
        hilos[1] = new Cliente("Luis", inst, 1, monitor);

        for (Thread hilo : hilos)
            hilo.start();

        for (Thread hilo : hilos)
            hilo.join();
        System.out.println("\n" + inst.toString());
    }
}

File: Prog. Concu y Paralela/Tarea 4/Monitores/TiendaDekkerMonitor.py
from threading import Thread, Condition, Lock
from typing import List
import random
import time

class Monitor:
    def __init__(self):
        self.turno = 0
        self.lock = Lock()
        self.condition = Condition(self.lock)
    
    def entrar_seccion(self, hilo_id: int):
        with self.lock:
            while self.turno != hilo_id:
                self.condition.wait()
    
    def salir_seccion(self, hilo_id: int):
        with self.lock:
            self.turno = 1 - hilo_id
            self.condition.notify_all()
                    

class Tienda:
    def __init__(self):
        self.stock = {
            "Camisa Azul chica": 2,
            "Pantalon verde mediano": 3,
            "Tennis negros chicos": 5,
        }
        self.prendas = list(self.stock.keys())

    def comprar(self, num_prenda):
        prenda = self.prendas[num_prenda]
        cantidad = self.stock[prenda]

        if cantidad == 0:
            return [prenda, None]

        self.stock[prenda] = cantidad - 1
        return [prenda, str(cantidad - 1)]

    def get_size(self):
        return len(self.stock)

    def __str__(self):
        output = ["Prenda\tCantidad disponible"]
        for prenda, cantidad in self.stock.items():
            output.append(f"{prenda}:\t{cantidad}")
        return "\n".join(output)


class Cliente(Thread):
    def __init__(self, client, hilo_id, tienda, ):
        super().__init__() #clase Thread de python
        self.tienda = tienda
        self.client = client
        self.hilo_id = hilo_id

    def run(self):
        random_num = random.randint(0, self.tienda.get_size() - 1)
        
        self.monitor.entrar(self.hilo_id)
        try:
            # SECCION CRITICA
            resultado = self.tienda.comprar(random_num)
            # SECCION CRITICA

            if resultado[1] is None:
                print(f"{self.client} quiso comprar" 
                      f"{resultado[0]} pero ya no hay existencias")
            else:
                print(
                    f"- {self.client} ha comprado: {resultado[0]} - "
                    f"Existencias restantes: {resultado[1]}, "
                    f"saliendo de la tienda..."
                )
        finally:
            self.monitor.salir(self.hilo_id)


class main():
    tienda = Tienda()
    print(tienda, "\n")
    
    hilos = [
        Cliente("Marcos", 1, tienda),
        Cliente("Lopez", 0, tienda),
    ]
    
    for hilo in hilos:
        hilo.start()
        
    for hilo in hilos:
        hilo.join()
        
    print(tienda, "\n")
    
    
main()